src/library/scala/reflect/ClassTag.scala

```diff
@@ -46,19 +46,7 @@ trait ClassTag[T] extends ClassManifestDeprecatedApis[T] with Equals with Serial
   def wrap: ClassTag[Array[T]] = ClassTag[Array[T]](arrayClass(runtimeClass))
 
   /** Produces a new array with element type `T` and length `len` */
-  override def newArray(len: Int): Array[T] =
-    runtimeClass match {
-      case java.lang.Byte.TYPE      => new Array[Byte](len).asInstanceOf[Array[T]]
-      case java.lang.Short.TYPE     => new Array[Short](len).asInstanceOf[Array[T]]
-      case java.lang.Character.TYPE => new Array[Char](len).asInstanceOf[Array[T]]
-      case java.lang.Integer.TYPE   => new Array[Int](len).asInstanceOf[Array[T]]
-      case java.lang.Long.TYPE      => new Array[Long](len).asInstanceOf[Array[T]]
-      case java.lang.Float.TYPE     => new Array[Float](len).asInstanceOf[Array[T]]
-      case java.lang.Double.TYPE    => new Array[Double](len).asInstanceOf[Array[T]]
-      case java.lang.Boolean.TYPE   => new Array[Boolean](len).asInstanceOf[Array[T]]
-      case java.lang.Void.TYPE      => new Array[Unit](len).asInstanceOf[Array[T]]
-      case _                        => java.lang.reflect.Array.newInstance(runtimeClass, len).asInstanceOf[Array[T]]
```

Comment:
Actually, the last line
```scala
java.lang.reflect.Array.newInstance(runtimeClass, len).asInstanceOf[Array[T]]
```
is a valid implementation for all the cases except `java.lang.Void.TYPE`. I suggest keeping this method concrete with just this implementation, and override it only in the `ClassTag` for `Unit` with
```scala
new Array[Unit](len).asInstanceOf[Array[T]]
```

Reply:
There is still a (small) performance benefit to the `newarray` over `.Array.newInstance(nonConstantArgument)`, so I'd be hesitant to make that change.